src/mad_spark_alt/evolution/cost_estimator.py

```
@@ -0,0 +1,658 @@
+    def get_usage_statistics(self) -> Dict[str, Any]:
+        """
+        Get statistics from historical usage.
+
+        Returns:
+            Usage statistics by operation type
+        """
+        if not self._historical_token_usage:
+            return {"message": "No historical data available"}
+
+        stats = {}
+        operation_types = set(u["operation_type"] for u in self._historical_token_usage)
```

Ruff (0.12.2)

Unnecessary generator (rewrite as a set comprehension)

Rewrite as a set comprehension

(C401)